test/extended/networking/multinetpolicy.go

```diff
@@ -0,0 +1,203 @@
+
+		nodes, err := e2enode.GetReadySchedulableNodes(f.ClientSet)
+		o.Expect(err).To(o.Succeed())
+		o.Expect(len(nodes.Items)).To(o.BeNumerically(">", 0))
+		scheduledNode := nodes.Items[0]
+
+		g.By("launching pods with an annotation to use the net-attach-def on node " + scheduledNode.Name)
+
+		frameworkpod.CreateExecPodOrFail(f.ClientSet, ns, podGenerateName, func(pod *v1.Pod) {
+			pod.Spec.NodeName = scheduledNode.Name
+			pod.Spec.Containers[0].Args = []string{"net", "--serve", podAListenAddress}
+			pod.ObjectMeta.Labels = map[string]string{"pod": "a"}
+			pod.ObjectMeta.Annotations = map[string]string{
+				"k8s.v1.cni.cncf.io/networks": fmt.Sprintf(networksTemplate, addrs.A.String())}
+		})
+
+		testPodB := frameworkpod.CreateExecPodOrFail(f.ClientSet, ns, podName, func(pod *v1.Pod) {
+			pod.Spec.NodeName = scheduledNode.Name
+			pod.ObjectMeta.Labels = map[string]string{"pod": "b"}
+			pod.ObjectMeta.Annotations = map[string]string{
+				"k8s.v1.cni.cncf.io/networks": fmt.Sprintf(networksTemplate, addrs.B.String())}
+		})
+
+		frameworkpod.CreateExecPodOrFail(f.ClientSet, ns, podName, func(pod *v1.Pod) {
+			pod.Spec.NodeName = scheduledNode.Name
+			pod.Spec.Containers[0].Args = []string{"net", "--serve", podCListenAddress}
+			pod.ObjectMeta.Labels = map[string]string{"pod": "c"}
+			pod.ObjectMeta.Annotations = map[string]string{
+				"k8s.v1.cni.cncf.io/networks": fmt.Sprintf(networksTemplate, addrs.C.String())}
+		})
+
+		g.By("checking podB can connect to podA")
+		podShouldReach(oc, testPodB.Name, podAListenAddress)
+
+		g.By("enabling MultiNetworkPolicies on cluster")
+		enableMultiNetworkPolicy(oc)
+
+		g.By("creating a deny-all-ingress traffic to pod:a MultiNetworkPolicy")
+		multinetpolicy_yaml := exutil.FixturePath("testdata", "multinetpolicy", "deny-ingress-pod-a.yml")
+		err = oc.AsAdmin().Run("create").Args("-f", multinetpolicy_yaml).Execute()
+		o.Expect(err).To(o.Succeed())
+
+		g.By("checking podB can NOT connect to podA")
+		podShouldNotReach(oc, testPodB.Name, podAListenAddress)
+
+		g.By("checking podB can still connect to podC")
+		podShouldReach(oc, testPodB.Name, podCListenAddress)
+	},
+		g.Entry("IPv4", podAddressSet{
+			A: mustParseIPAndMask("192.0.2.1/24"),
+			B: mustParseIPAndMask("192.0.2.2/24"),
+			C: mustParseIPAndMask("192.0.2.3/24"),
+		}),
+		g.Entry("IPv6", podAddressSet{
+			A: mustParseIPAndMask("2001:db8::1/32"),
+			B: mustParseIPAndMask("2001:db8::2/32"),
+			C: mustParseIPAndMask("2001:db8::3/32"),
+		}),
+	)
+
+})
+
+func enableMultiNetworkPolicy(oc *exutil.CLI) {
+	c := oc.AdminOperatorClient().OperatorV1().Networks()
+	config := getCluserNetwork(c)
+
+	b := true
+	config.Spec.UseMultiNetworkPolicy = &b
+
+	_, err := c.Update(context.Background(), config, metav1.UpdateOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	o.Eventually(func() error {
+		return oc.AsAdmin().Run("get").Args("multi-networkpolicies.k8s.cni.cncf.io").Execute()
+	}, "30s", "2s").Should(o.Succeed())
```

Comment:
How about to check daemonset, multus-networkpolicy, instead of CRD, to capture multi-networkpolicy-iptables Pod failure before actual tests?

Reply:
I believe we need to check the daemonset readiness here for further troubleshooting.

Let's imagine we have a trouble in multi-networkpolicy-iptables Pod (like crashed and never goes to 'Running' state). In such case, current test just files error at 'traffic testing': 'podShoudNotReach() is failed'. Usually if we get 'podShouldNotReach() is failed' then we suspect that policy rule/iptables functionality. So it may be hard to proceed to troubleshoot.

If we identify the pod crash here, then easy to understand what is happen, so could you please check DaemonSet readiness, such like that?

https://github.com/openshift/origin/blob/9f98fce6e2932bea2cf28205b3eb7359222b7ec2/test/extended/dr/common.go#L204-L208

Reply:
Got it! Thanks for pointing out